core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -164,6 +170,11 @@ object Channel {
                       State(a :: values, size + 1, None, producers, false),
                       notifyStream(waiting).as(rightUnit)
                     )
+                  else if (close)
+                    (
+                      State(values, size, None, producers, true),
+                      notifyStream(waiting).as(rightUnit) <* signalClosure
+                    )
                   else
                     (
                       State(values, size, None, (a, producer) :: producers, false),
```

Comment:
Hmm, this logic doesn't seem right to me. Whether `size < capacity` or not is indepndent of `close`. For example, if `size < capacity` then you take the first branch and are not closing the channel, which is incorrect.

I think it should be something like this.

```scala
                  if (size < capacity)
                    (
                      State(a :: values, size + 1, None, producers, close),
                      signalClosure.whenA(close) *> notifyStream(waiting).as(rightUnit)
                    )
                  else
                    (
                      State(values, size, None, (a, producer) :: producers, close),
                      signalClosure.whenA(close) *> notifyStream(waiting).as(rightUnit) <* waitOnBound(producer, poll)
                    )
```

Reply:
Yes, indeed. This makes sense! Thanks!